src/hotspot/cpu/aarch64/aarch64_neon_ad.m4

```diff
@@ -2243,3 +2243,140 @@ instruct vpopcount$1$2`'(vec$5 dst, vec$5 src) %{
 dnl       $1 $2 $3  $4 $5
 VPOPCOUNT(4, I, 16, 8, X)
 VPOPCOUNT(2, I, 8,  4, D)
+
+dnl
+define(`VMASK_TRUECOUNT', `
+instruct vmask_truecount$1(iRegINoSp dst, $2 src, $2 tmp) %{
+  predicate(n->in(1)->bottom_type()->is_vect()->element_basic_type() == T_BOOLEAN);
+  match(Set dst (VectorMaskTrueCount src));
+  effect(TEMP tmp);
+  ins_cost(2 * INSN_COST);
+  format %{ "addv $tmp, $src\n\t"
+            "umov $dst, $tmp, B, 0\t# vector ($1)" %}
+  ins_encode %{
+    __ addv(as_FloatRegister($tmp$$reg), __ T$1, as_FloatRegister($src$$reg));
+    __ umov($dst$$Register, as_FloatRegister($tmp$$reg), __ B, 0);
+  %}
+  ins_pipe(pipe_slow);
+%}')dnl
+dnl
+// vector mask reductions
+dnl             $1   $2
+VMASK_TRUECOUNT(8B,  vecD)
+VMASK_TRUECOUNT(16B, vecX)
+
+instruct vmask_firsttrue_LT8B(iRegINoSp dst, vecD src, vecD tmp, rFlagsReg cr) %{
+  predicate(n->in(1)->bottom_type()->is_vect()->element_basic_type() == T_BOOLEAN &&
+            n->in(1)->bottom_type()->is_vect()->length() < 8);
+  match(Set dst (VectorMaskFirstTrue src));
+  effect(TEMP_DEF dst, TEMP tmp, KILL cr);
+  ins_cost(8 * INSN_COST);
+  format %{ "vmask_firsttrue $dst, $src\t# vector (4I/4S/2I)" %}
+  ins_encode %{
+    // Revert the bits and count the leading zero bytes. Get the maximum value
+    // between the vector length and the leading zero bytes.
+    __ negr(as_FloatRegister($tmp$$reg), __ T8B, as_FloatRegister($src$$reg));
+    __ fmovd($dst$$Register, as_FloatRegister($tmp$$reg));
+    __ rbit($dst$$Register, $dst$$Register);
+    __ clz($dst$$Register, $dst$$Register);
+    __ lsrw($dst$$Register, $dst$$Register, 3);
+    __ movw(rscratch1, vector_length(this, $src));
+    __ cmpw($dst$$Register, rscratch1);
+    __ cselw($dst$$Register, rscratch1, $dst$$Register, Assembler::GE);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vmask_firsttrue8B(iRegINoSp dst, vecD src, vecD tmp) %{
+  predicate(n->in(1)->bottom_type()->is_vect()->element_basic_type() == T_BOOLEAN &&
+            n->in(1)->bottom_type()->is_vect()->length() == 8);
+  match(Set dst (VectorMaskFirstTrue src));
+  effect(TEMP_DEF dst, TEMP tmp);
+  ins_cost(5 * INSN_COST);
+  format %{ "vmask_firsttrue $dst, $src\t# vector (8B)" %}
+  ins_encode %{
+    // Revert the bits and count the leading zero bytes.
+    __ negr(as_FloatRegister($tmp$$reg), __ T8B, as_FloatRegister($src$$reg));
```

Comment:
Should that be "Reverse the bits?" But in any case, we can see that the code calls rbit then clz, presumably because you want to count the trailing bits. What does the negr do here?

Reply:
Thanks for your comment @theRealAph !  Yeah, it's "Reverse the bits".  Thanks for pointing out this!
`negr` is used to convert the original mask values from `0/1` to `0/-1`. So that we can firstly calculate the trailing zero numbers in bits, and then divide it by 8. Note that the mask input for these operations are different from others. It is actually a boolean vector with `0/1` as the element value.

Reply:
Yes, but you reverse the bits in $dst, which reverses the byte order and changes each 0x01 byte to 0x80. You then count the leading zeroes. What, then, is the point of negating each 0x01 byte to 0xff first?

Reply:
Yes, I was intending to make the bits right after reversing. But seems it's also ok to remove the `negr` for `firsttrue`. And for `lasttrue`, I think it's better to keep the `negr`. I will remove it for `firsttrue` and have a test. Thanks for the review!

Reply:
Seems `negr` is also needless for `lasttrue`, since `7 >> 3 == 0` ?

Reply:
> Seems `negr` is also needless for `lasttrue`, since `7 >> 3 == 0` ?

Yeah. The in-reg mask representation of these VectorMaskOp nodes (0/1) are different from anytrue/alltrue (VectorTest, 0/-1), which looks a bit confusing.

Reply:
Hi @theRealAph , the new commit has removed the redundant `negr` for `firsttrue, lasttrue`, and all tests pass. Would you mind having look at it again? Thanks so much!

Reply:
> > Seems `negr` is also needless for `lasttrue`, since `7 >> 3 == 0` ?
> 
> Yeah. And the in-reg mask representation of these VectorMaskOp nodes (0/1) are different from anytrue/alltrue (VectorTest, 0/-1), which looks a bit confusing.

Yes, to make it clear, I'v added some comments for it. Thanks!